File: src/rotation.py
"Rotate matrices!"


def out_of_place(arr):
    l = len(arr)
    output = [[None for _ in range(l)] for _ in range(l)]

    for x, row in enumerate(arr):
        for y, cell in enumerate(row):
            output[y][l - 1 - x] = cell

    return output


def in_place(arr):
    l = len(arr)
    o = l - 1

    for x in range(l):
        for y in range(0, l - x):
            arr[x][y], arr[o - y][o - x] = arr[o - y][o - x], arr[x][y]

    for i in range(l // 2):
        arr[i], arr[o - i] = arr[o - i], arr[i]

    return arr

File: src/rotation.js
const outOfPlace = arr =>
  arr.map((row, x) => row.map((cell, y) => arr[arr.length - 1 - y][x]));

const inPlace = arr => {
  const l = arr.length;
  const o = l - 1;

  for (let x = 0; x < l; x++) {
    for (let y = 0; y < l - x; y++) {
      [arr[x][y], arr[o - y][o - x]] = [arr[o - y][o - x], arr[x][y]];
    }
  }

  for (let i = 0; i < Math.floor(l / 2); i++) {
    [arr[i], arr[o - i]] = [arr[o - i], arr[i]];
  }

  return arr;
};
module.exports = { outOfPlace, inPlace };
